File: old_trunk/cost/iona/orbacus/java/poa/TestLocationForwardClient.java
// **********************************************************************
//
// Copyright (c) 2001
// Object Oriented Concepts, Inc.
// Billerica, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

package test.poa;

import java.util.*;
import java.io.*;
import org.omg.CORBA.*;
import org.omg.PortableServer.*;
import org.omg.PortableServer.POAPackage.*;

final public class TestLocationForwardClient extends test.common.TestBase
{
    public static void
    main(String args[])
    {
        Properties props = System.getProperties();

        ORB orb = null;

	try
	{
            //
            // Create ORB
            //
	    orb = ORB.init(args, props);

	    POA root = TestUtil.GetRootPOA(orb);

	    POA poa;
	    Policy[] policies;

	    POAManager manager = root.the_POAManager();

	    //
	    // Create POA
	    //
	    policies = new Policy[4];
	    policies[0] = root.create_id_assignment_policy(
		org.omg.PortableServer.IdAssignmentPolicyValue.USER_ID);
	    policies[1] = root.create_id_uniqueness_policy(
		org.omg.PortableServer.IdUniquenessPolicyValue.UNIQUE_ID);
	    policies[2] = root.create_servant_retention_policy(
		org.omg.PortableServer.ServantRetentionPolicyValue.RETAIN);
	    policies[3] = root.create_request_processing_policy(
		org.omg.PortableServer.RequestProcessingPolicyValue.
		USE_SERVANT_MANAGER);
	    try
	    {
		poa = root.create_POA("poa", manager, policies);
	    }
	    catch(InvalidPolicy ex)
	    {
		throw new RuntimeException();
	    }
	    catch(AdapterAlreadyExists ex)
	    {
		throw new RuntimeException();
	    }

	    TestLocationForwardActivator_impl activatorImpl =
		new TestLocationForwardActivator_impl();
	    ServantActivator activator = activatorImpl._this(orb);
	    try
	    {
		poa.set_servant_manager(activator);
	    }
	    catch(WrongPolicy ex)
	    {
		throw new RuntimeException();
	    }

	    byte[] oid = ("test").getBytes();
            org.omg.CORBA.Object reference =
                poa.create_reference_with_id(oid, "IDL:Test:1.0");

	    String impl = null;

	    //
	    // Read all object references from file
	    //
	    try
	    {
		String refFile = "Test.ref";
		FileInputStream file = new FileInputStream(refFile);
		BufferedReader in =
		    new BufferedReader(new InputStreamReader(file));
		impl = in.readLine();
		file.close();
	    }
	    catch(IOException ex)
	    {
		System.err.println("Can't read from `" +
				   ex.getMessage() + "'");
		System.exit(1);
	    }

	    org.omg.CORBA.Object obj = orb.string_to_object(impl);
	    TestLocationForwardServer server =
		TestLocationForwardServerHelper.narrow(obj);
	    TEST(server != null);
	    org.omg.CORBA.Object servant = server.get_servant();

	    activatorImpl.setForwardRequest(servant);

	    TestLocationForward_impl testImpl =
		new TestLocationForward_impl(orb);
	    activatorImpl.setActivatedServant(testImpl);

	    try
	    {
		manager.activate();
	    }
	    catch( org.omg.PortableServer.POAManagerPackage.AdapterInactive ex)
	    {
		throw new RuntimeException();
	    }

	    server.setForwardRequest(reference);

	    //
	    // Run some calls
	    //
	    TestLocationForward local =
		TestLocationForwardHelper.narrow(reference);

	    //
	    // First should be local
	    //
	    local.aMethod();
	    local.deactivate_servant();

	    //
	    // Second, should be remote
	    //
	    local.aMethod();
	    local.deactivate_servant();

	    //
	    // Third should be local again
	    //
	    local.aMethod();
	    local.deactivate_servant();

	    //
	    // Clean up
	    //
	    poa.destroy(true, true);

	    server.deactivate();
	}
	catch(SystemException ex)
	{
            ex.printStackTrace();
            System.exit(1);
	}

        if(orb != null)
        {
            try
            {
                orb.destroy();
            }
            catch(SystemException ex)
            {
                ex.printStackTrace();
                System.exit(1);
            }
        }

	System.exit(0);
    }
}

File: cost/iona/orbacus/cpp/poa/TestLocationForwardClient.cpp
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <includes.h>

#include <TestUtil.h>

#include <TestLocationForwardActivator_skel.h>

#include <TestLocationForwardActivator_impl.h>
#include <Test_impl.h>
#include <TestLocationForward_impl.h>

#include <fstream>

using namespace CORBA;
using namespace PortableServer;

int
main(int argc, char* argv[], char*[])
{
    int status = EXIT_SUCCESS;
    ORB_var orb;

    try
    {
        orb = ORB_init(argc, argv);

        POA_var root = TestUtil_GetRootPOA(orb);

        POA_var poa;
        PolicyList policies;

        POAManager_var manager = root -> the_POAManager();
        
        //
        // Create POA
        //
        policies.length(4);
        policies[0] =
            root -> create_id_assignment_policy(PortableServer::USER_ID);
        policies[1] =
            root -> create_id_uniqueness_policy(PortableServer::UNIQUE_ID);
        policies[2] =
            root -> create_servant_retention_policy(PortableServer::RETAIN);
        policies[3] = root ->
         create_request_processing_policy(PortableServer::USE_SERVANT_MANAGER);
        poa = root -> create_POA("poa", manager, policies);

        TestLocationForwardActivator_impl* activatorImpl = 
	    new TestLocationForwardActivator_impl();
	PortableServer::ServantBase_var activatorServant = activatorImpl;
        PortableServer::ServantActivator_var activator =
	    activatorImpl -> _this();
        poa -> set_servant_manager(activator);

        ObjectId_var oid = string_to_ObjectId("test");
        Object_var reference =
            poa -> create_reference_with_id(oid, "IDL:Test:1.0");

        //
        // Read all object references from file
        //
        const char* refFile = "Test.ref";
        ifstream in; // Must use open(name), not ifstream in(name) (VC++ bug)
        in.open(refFile);
        if(in.fail())
        {
            OB_ERROR(argv[0] << ": can't open `" << refFile << "'");
            return EXIT_FAILURE;
        }

        char s[1000];
        in >> s;
	in.close();

        Object_var obj = orb -> string_to_object(s);
        if(CORBA::is_nil(obj))
        {
            OB_ERROR(argv[0] << ": cannot read IOR from Test.ref");
            return EXIT_FAILURE;
        }
        TestLocationForwardServer_var server =
            TestLocationForwardServer::_narrow(obj);
        TEST(!is_nil(server));
        Object_var servant = server -> get_servant();

	activatorImpl -> setForwardRequest(servant);

	TestLocationForward_impl* testImpl =
	    new TestLocationForward_impl(orb, root);
        PortableServer::ServantBase_var testServant = testImpl;
        activatorImpl -> setActivatedServant(testServant);

        manager -> activate();

	server -> setForwardRequest(reference);

        //
        // Run some calls
        //
        TestLocationForward_var local =
            TestLocationForward::_narrow(reference);

        //
        // First should be local
        //
        local -> aMethod();
        local -> deactivate_servant();

        //
        // Second, should be remote
        //
        local -> aMethod();
        local -> deactivate_servant();

        //
        // Third should be local again
        //
        local -> aMethod();
        local -> deactivate_servant();

        //
        // Clean up
        //
        poa -> destroy(true, true);

	server -> deactivate();
    }
    catch(const Exception& ex)
    {
#ifdef HAVE_EXCEPTION_INSERTERS
	OB_ERROR(ex);
#else
	cerr << "Exception: " << ex._rep_id() << endl;
#endif
        status = EXIT_FAILURE;
    }

    if(!is_nil(orb))
    {
	try
	{
	    orb -> destroy();
	}
	catch(const Exception& ex)
	{
#ifdef HAVE_EXCEPTION_INSERTERS
	    OB_ERROR(ex);
#else
	    cerr << "Exception: " << ex._rep_id() << endl;
#endif
	    status = EXIT_FAILURE;
	}
    }

    return status;
}
